src/renderer/dataframe/Dataframe.js

```diff
@@ -0,0 +1,322 @@
+import { GEOMETRY_TYPE } from '../../utils/geometry';
+import DummyDataframe from './DummyDataframe';
+import { WM_R } from '../../utils/util';
+
+import { genViewportFeatureClass } from './viewportFeature';
+
+import FeatureIdsHelper from './FeatureIdsHelper';
+import ViewportHelper from './ViewportHelper';
+
+// Maximum number of property textures that will be uploaded automatically to the GPU
+// in a non-lazy manner
+const MAX_GPU_AUTO_UPLOAD_TEXTURE_LIMIT = 32;
+
+const featureClassCache = new Map();
+
+export default class Dataframe extends DummyDataframe {
+    constructor (...args) {
+        super(...args);
+
+        this._initHelpers();
+    }
+
+    _initHelpers () {
+        if (!this._featureIdsHelper) {
+            this._featureIdsHelper = new FeatureIdsHelper(this);
+        }
+
+        if (!this._viewportHelper) {
+            this._viewportHelper = new ViewportHelper(this);
+        }
+    }
+
+    /**
+     * Loads all the relevant information to the WebGL context, provided by the renderer.
+     * This includes:
+     *   - the textures for each property
+     *   - the buffers for vertices & normals
+     *   - the auxiliary textures for style properties
+     *
+     * This method is very intensive in terms of computations, loading potentially a lot of CPU data to the GPU
+     *
+     */
+    bindRenderer (renderer) {
+        this.renderer = renderer;
+        this.height = this.getSize().height;
```

Comment:
Maybe we can remove this line, and also the other assignment to update the dataframe height? I _think_ it's no longer needed anywhere

Reply:
I think I have found an external use. See next commit for a little refactor to remove it